### DISCUSSION TITLE
[Fooocus 2.2.0] Metadata, Isolation, Array Support, Docker

### DISCUSSION THREAD

**Discussion by mashb1t** [lllyasviel/Fooocus#2410] [General]
## TL;DR
* Isolate every image generation to truly allow multi-user usage
* Add array support, changes the main prompt when increasing the image number. Syntax: `[[red, green, blue]] flower` 
* Add optional metadata to images, allowing you to regenerate and modify them later with the same parameters
* Now supports native PNG, JPG and WEBP image generation
* Add Docker support

## What's Changed
* fix: sort with casefold, case insensitive by @mashb1t
* feat: add early return for prompt expansion when no new tokens should be added by @mashb1t
* feat: ignore DS_Store by @charliewilco in https://github.com/lllyasviel/Fooocus/pull/2313
* feat: advanced params refactoring + prevent users from skipping/stopping other users tasks in queue by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/981
* feat: add list of 100 most popular animals to wildcards by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/985
* feat: add advanced parameter for disable_intermediate_results (progress_gallery) by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1013
* feat: add ability to load checkpoints and loras from multiple locations by @dooglewoogle in https://github.com/lllyasviel/Fooocus/pull/1256
* feat: allow users to specify the number of threads when running on CPU by @maxim-saplin in https://github.com/lllyasviel/Fooocus/pull/1601
* feat: improve bug report and feature request issue templates by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1631
* fix: correctly create directory for path_outputs if not existing by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1668
* fix: allow path_outputs to be outside of root dir by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2332
* feat: add button to enable LoRAs by @MindOfMatter in https://github.com/lllyasviel/Fooocus/pull/2210
* feat: make lora number editable in config by @MindOfMatter in https://github.com/lllyasviel/Fooocus/pull/2215
* feat: make lora min max weight editable in config by @MindOfMatter in https://github.com/lllyasviel/Fooocus/pull/2216
* feat: add array support on main prompt by @flannerybh in https://github.com/lllyasviel/Fooocus/pull/1503
* feat: use consistent file name in gradio by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1932
* feat: add metadata to images by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1940
* feat: add jpg and webp support, add exif data handling for metadata by @mashb1t and @eddyizm in https://github.com/lllyasviel/Fooocus/pull/1863
* feat: add docker files by @whitehara and @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1418
* docs: fix typo in readme by @gteti in https://github.com/lllyasviel/Fooocus/pull/2368

## New Contributors
* @charliewilco made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2313
* @dooglewoogle made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1256
* @maxim-saplin made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1601
* @MindOfMatter made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2210
* @flannerybh made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1503
* @whitehara made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1418
* @gteti made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2368

**Full Changelog**: https://github.com/lllyasviel/Fooocus/compare/2.1.865...2.2.0

**Comment by poor7** [comment]
Syntax:` [[red, green, blue]] flower` He always draws in the same sequence, first red, then green, and then a blue flower and again. Is this how it should be?

**Reply by flannerybh** [reply]
Yes, that is the intended result. It's like a wildcard, but without the randomness.

**Reply by mashb1t** [reply]
@poor7 ~~this is why i wrote "when increasing the image number", which you have to do to get images according to all entries of your array. Please increase the image number and check the output.~~ NVM, misunderstood your comment.
@flannerybh thanks again for the implementation!

**Comment by bsdesigner** [comment]
"Generate Image Grid for Each Batch" does not work (((
![изображение_2024-03-03_003111001](https://github.com/lllyasviel/Fooocus/assets/36730003/243eb44b-9ff1-4f83-be65-321188cd3bbf)

**Reply by mashb1t** [reply]
This went under the radar, thank you for reporting. Fixed in https://github.com/lllyasviel/Fooocus/pull/2414, please restart Fooocus and check again.

**Reply by bsdesigner** [reply]
It's all working!!! I'm working in Fooocus via Stability Matrix, so I updated a little late. This is a very useful feature when testing! Thank you very much for your promptness!!!

**Comment by Brassee** [comment]
Images no longer saves in 'outputs' 

UPD. I found new folder 'outputs' in user 'download' folder. All freshly generated images are there

**Reply by mashb1t** [reply]
Please check the path_outputs in the config.txt file and the terminal output for irregularities on application start or error messages. If you think there is something wrong, please gather all information and open an issue.

**Reply by Brassee** [reply]
Everything seems ok now Thanks!

**Comment by patrickchamp** [comment]
Love the update.

I’m trying to cycle through wildcards or arrays using the same seed. I thought disable_seed_increment might do it but the images are all identical. Can you clarify what this checkbox does? I think I just read that “Generate forever” on a batch of 1 might cycle through wildcards randomly but can’t use it atm because I’m on an iPad.

**Reply by patrickchamp** [reply]
Update: It was ignoring the array because one of the values had parentheses. I removed them and it’s properly working for arrays. Ideally, the seed increment would have the same effect on wildcards. In order to test values against the same seed, I had to take a wildcard list and convert it into a comma separated list.

**Reply by mashb1t** [reply]
@freshp124 wildcard read in order has not yet been merged, but is in this PR: https://github.com/lllyasviel/Fooocus/pull/1761

**Comment by saadkhan955** [comment]
Is there a way to use multiple arrays? For example, could I create an array like `A [[chair, table, lamp, bed]] and another array  [[wood, iron, steel, plastic]]` to generate the image? I tried a method like this, but it iterates through the first array and only selects "wood" from the second array.

**Reply by flannerybh** [reply]
That should work, can you post your full prompt? Remember to set the number of images to 16

**Reply by saadkhan955** [reply]
Thanks, it's working now after setting the image count to 16. I'm wondering if there's a way to select only one property from each array for each image. For example, on the first run, the prompt should be "A chair made out of wood," and on the second run, the prompt should be "A table made out of iron."

**Reply by flannerybh** [reply]
[[ A chair made out of wood, A table made out of iron ]]

**Comment by poor7** [comment]
How to load metadata from a saved image back into Fooocus?

**Reply by mashb1t** [reply]
Input Image > Metadata > Apply Metadata

**Comment by poor7** [comment]
If selected **Output Format**=**jpg** or **webp** then **Save Metadata to Images**=**a1111 (plain text)** does not work.

**Reply by mashb1t** [reply]
No issues for me, are there any error logs? Which program do you use to read metadata? If you don't use Fooocus (Input Image > Metadata), make sure it can read EXIF data.

**Reply by poor7** [reply]
a1111 - PNG Info

**Reply by mashb1t** [reply]
> Output Format=jpg or webp

> PNG Info

Think about it for a second :D
JPG and WebP both use EXIF to save and read metadata as only PNG has PngInfo. The Fooocus Metadata implementation does support all 3 file formats as well as EXIF and PngInfo saving/reading. Please find it in Input Image > Metadata.

**Reply by eddyizm** [reply]
So I am either missing something or misunderstanding. I tested this and am able to get jpg's out with the metadata and is read by fooocus as expected, 

However Civitia.com does not see any of the metadata, which I expected it to. 
![image](https://github.com/lllyasviel/Fooocus/assets/12236892/40971943-070c-4a5b-977c-61e2ec7c3b29)

![image](https://github.com/lllyasviel/Fooocus/assets/12236892/08117411-4146-4ea1-8b7f-ceb099c47d58)

Am I missing a step or do I have a misunderstanding?

**Reply by mashb1t** [reply]
Sadly only png's in a1111 scheme are compatible with the parsing of Civitai.
JPG and WebP use exif for parsing, which is supported in Fooocus, but not other well-known AI platforms.
Maybe we should add a hint stating sth. like "Use png and a1111 [...]"

**Reply by eddyizm** [reply]
lol holy crap. I always use PNG and i thought for this feature the jpg would be ideal since I am very familiar with exif (photographer)  so I literally shot myself in the foot and was totally wrong about it. 

Ok so maybe I will put in a small pr to put that note in the label?

**Reply by mashb1t** [reply]
Sure, feel free, would do it tomorrow anyways!
Adjustments in 2 places, info and en.json

**Reply by eddyizm** [reply]
ok done. 
https://github.com/lllyasviel/Fooocus/pull/2434

**Reply by eddyizm** [reply]
Another question on the metadata, I added about 5 loras to test the civitia integration and although feeding it back to fooocus there is data and hashes regarding the loras used, civitia only sees the base checkpoint and no lora info at all. 

Just want to verify if this is the expected outcome.

**Reply by mashb1t** [reply]
Yes, this is expected. Civitai can not read LoRAs from A1111 scheme, even when generated with A1111 (just checked with v1.8.0).

![image](https://github.com/lllyasviel/Fooocus/assets/9307310/8a6a071b-c81f-4359-a96e-04c14462499c)

**Comment by eddyizm** [comment]
Interestingly this morning I was testing it out and found something a little odd. 
@mashb1t I believe you have mentioned elsewhere to backup the `config.txt` elsewhere 
I loaded the update to a new machine with an existing install and gave a few lora a try with some of the new features enabled. 
The output was a little odd so when I checked the log, it seemed to have enabled two Fooocus styles, which in my preset are disabled. 
`"default_styles": [],`
Nothing that I can see in the config though so still tracking down how this happened. Gonna spin up a fresh session on google colab to see if I can reproduce it.

**Reply by eddyizm** [reply]
ok, after a few experiments, It looks like these values get set when using the `describe` feature, unbeknownst to me, 

https://github.com/lllyasviel/Fooocus/blob/4ea3baff501f876f1e1e6628491b56fb5a3832ee/webui.py#L659

A little alarming when I wasn't expecting it but apparently not really a bug. I have not really ventured too deep into all the levers and buttons Fooocus has to offer.

**Comment by mashb1t** [comment]
2.2.1 is planned to be released tomorrow, content: https://github.com/lllyasviel/Fooocus/milestone/2?closed=1

**Reply by mashb1t** [reply]
=> https://github.com/lllyasviel/Fooocus/releases/tag/2.2.1 / https://github.com/lllyasviel/Fooocus/discussions/2440

**Comment by jaimeastorga2000** [comment]
The last two releases have not had binaries, only source code. I know existing installs auto-update if you have an internet connection, but can we please get the binaries as well? I know I would have never tried Fooocus if I had only seen source code on the release page.

**Reply by mashb1t** [reply]
The 7z file will be updated on package updates. As the last few releases didn't have any, there are no 7z files attached. FYI each 7z file is 1.7GB+ in size.

As you've mentioned correctly, you can simply download the code and use the auto-update or download the 7z file from the readme + extract and run it. After updating you can bundle the code again if you need it bundled in the latest version.

**Comment by eddyizm** [comment]
So not sure if this is already addressed some where but I have been seeing this error in my log a few times. 
`generate_image_grid = task.args.pop(0) IndexError: pop from empty list`

**Reply by mashb1t** [reply]
This has not been reported and is also working fine in main and develop as well as my and your fork. Please create an issue if needed.

**Reply by andynormancx** [reply]
I've also started getting this, happened twice today so far. I've had to restart fooocus when it happens.

**Comment by poor7** [comment]
Is it possible to add additional metadata directly to the file properties, in particular in the “Comments” section, so that you can view them without using Fooocus? It would be much more convenient than now.

**Reply by mashb1t** [reply]
please open a new discussion or feature request with more information about what you're suggesting.
